File: services/payment/src/main/kotlin/com/codejsha/bookstore/payment/domain/model/PaymentMapper.kt
package com.codejsha.bookstore.payment.domain.model

import com.codejsha.bookstore.service.application.port.openapi.model.PaymentCreateWebReq
import com.codejsha.bookstore.service.application.port.openapi.model.PaymentFindAllWebParam
import com.codejsha.bookstore.service.application.port.openapi.model.PaymentUpdateWebReq
import com.codejsha.bookstore.service.application.port.pb.paymentpb.PaymentFindAllProtoReq

sealed class PaymentMapper {
    companion object {
        fun toPaymentDto(req: PaymentCreateWebReq): PaymentDto = req.toPaymentDto()
        fun toPaymentDto(id: Long, req: PaymentUpdateWebReq): PaymentDto = req.toPaymentDto(id)

        fun toFilterCondition(param: PaymentFindAllWebParam): FilterCondition = param.toFilterCondition()
        fun toFilterCondition(request: PaymentFindAllProtoReq): FilterCondition = request.toFilterCondition()
    }
}

private fun PaymentCreateWebReq.toPaymentDto(): PaymentDto {
    return PaymentDto(
        id = null,
        orderId = this.orderId,
        userId = this.userId,
        paymentType = this.paymentType,
        cardNumber = this.cardNumber,
        amount = this.amount,
        paymentDate = this.paymentDate?.toLocalDateTime()
    )
}

private fun PaymentUpdateWebReq.toPaymentDto(id: Long): PaymentDto {
    return PaymentDto(
        id = id,
        orderId = this.orderId,
        userId = this.userId,
        paymentType = this.paymentType,
        cardNumber = this.cardNumber,
        amount = this.amount,
        paymentDate = this.paymentDate?.toLocalDateTime()
    )
}

private fun PaymentFindAllWebParam.toFilterCondition(): FilterCondition {
    return FilterCondition(
        filter = FilterCondition.QueryFilter(
            sort = this.sort,
            order = this.order,
            limit = this.limit,
            offset = this.offset
        )
    )
}

private fun PaymentFindAllProtoReq.toFilterCondition(): FilterCondition {
    return FilterCondition(
        filter = FilterCondition.QueryFilter(
            sort = this.filter.sort,
            order = this.filter.order,
            limit = this.filter.limit,
            offset = this.filter.offset
        )
    )
}

File: services/customer/internal/domain/model/mapper/payment_mapper.go
package mapper

import (
	"fmt"

	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/openapi"
	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/pb/paymentpb"
)

func NewPaymentMapper() *PaymentMapper {
	return &PaymentMapper{}
}

type PaymentMapper struct {
}

func (m PaymentMapper) ToPaymentFindAllProtoReq(value interface{}) (*paymentpb.PaymentFindAllProtoReq, error) {
	switch value.(type) {
	case openapi.PaymentFindAllWebParam:
		v := value.(openapi.PaymentFindAllWebParam)
		request := &paymentpb.PaymentFindAllProtoReq{
			UserId:  &v.Id,
			OrderId: v.OrderId,
			Filter: &paymentpb.QueryFilter{
				Sort:   v.Sort,
				Order:  v.Order,
				Limit:  v.Limit,
				Offset: v.Offset,
			},
		}
		return request, nil
	default:
		return nil, fmt.Errorf("failed to convert request to PaymentFindAllRequest: %v", value)
	}
}

func (m PaymentMapper) ToPaymentFindProtoReq(value interface{}) (*paymentpb.PaymentFindProtoReq, error) {
	switch value.(type) {
	case openapi.PaymentFindWebParam:
		v := value.(openapi.PaymentFindWebParam)
		request := &paymentpb.PaymentFindProtoReq{
			Id: v.PaymentId,
		}
		return request, nil
	default:
		return nil, fmt.Errorf("failed to convert request to PaymentFindRequest: %v", value)
	}
}
